src/metabase/models/card.clj

```diff
@@ -344,8 +361,13 @@
       ;; make sure this Card doesn't have circular source query references if we're updating the query
       (when (:dataset_query changes)
         (check-for-circular-source-query-references changes))
+      ;; updating a model dataset query to includes any clauses will disable implicit acitons(if it exists)
+      (when (and (:dataset_query changes)
+                 (:dataset old-card-info)
+                 (not (model-supports-implicit-actions? changes)))
+        (disable-implicit-action-for-model! id))
       ;; Archive associated actions
-      (when (and (not (:dataset changes))
+      (when (and (false? (:dataset changes))
                  (:dataset old-card-info))
         (t2/update! 'Action {:model_id id} {:archived true}))
```

Comment:
Not related to your PR but I just realised this is a bug too. Implicit actions should be deleted when a card is changed from a model to a card. Currently they are archived. Only query actions should be archived.

Reply:
nice catch. you'll fix it, right?

Reply:
yes